modules/google-analytics/wp-google-analytics.php

```diff
@@ -0,0 +1,408 @@
+		}
+
+		$track = array();
+		if ( is_404() && ( ! isset( $wga['log_404s'] ) || 'false' !== $wga['log_404s'] ) ) {
+			// This is a 404 and we are supposed to track them.
+			$custom_vars[] = "_gaq.push( [ '_trackEvent', '404', document.location.href, document.referrer ] );";
+		} elseif ( is_search() && ( ! isset( $wga['log_searches'] ) || 'false' !== $wga['log_searches'] ) ) {
+			// Set track for searches, if it's a search, and we are supposed to.
+			$track['data'] = sanitize_text_field( wp_unslash( $_REQUEST['s'] ) ); // Input var okay.
+			$track['code'] = 'search';
+		}
+
+		if ( ! empty( $track ) ) {
+			$track['url'] = $this->_get_url( $track );
+			// adjust the code that we output, account for both types of tracking.
+			$track['url'] = esc_js( str_replace( '&', '&amp;', $track['url'] ) );
+			$custom_vars[] = "_gaq.push(['_trackPageview','{$track['url']}']);";
+		} else {
+			$custom_vars[] = "_gaq.push(['_trackPageview']);";
+		}
+
+		if ( 'true' === $wga['enable_display_advertising'] ) {
+			$async_code = "<script type='text/javascript'>
+								var _gaq = _gaq || [];
+								%custom_vars%
+
+								(function() {
+									var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
+									ga.src = ('https:' == document.location.protocol ? 'https://' : 'http://') + 'stats.g.doubleclick.net/dc.js';
+									var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
+								})();
+							</script>";
+		} else {
+			$async_code = "<script type='text/javascript'>
+								var _gaq = _gaq || [];
+								%custom_vars%
+
+								(function() {
+									var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
+									ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
+									var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
+								})();
+							</script>";
+		}
+
+		$custom_vars_string = implode( "\r\n", $custom_vars );
+		$async_code = str_replace( '%custom_vars%', $custom_vars_string, $async_code );
+
+		return $this->_output_or_return( $async_code, $output );
+
+	}
+
+	/**
+	 * Used to get one or all of our plugin options
+	 *
+	 * @param string[optional] $option - Name of options you want.  Do not use if you want ALL options.
+	 * @param boolean[optiona] $default - Default value.
+	 * @return array of options, or option value.
+	 */
+	private function _get_options( $option = null, $default = false ) {
+
+		$o = get_option( 'jetpack_wga' );
+
+		if ( isset( $option ) ) {
+			if ( isset( $o[ $option ] ) ) {
+				if ( 'code' === $option ) {
+					if ( preg_match( '#UA-[\d-]+#', $o[ $option ], $matches ) ) {
+						return $matches[0];
+					} else {
+						return '';
+					}
+				} else {
+					return $o[ $option ];
+				}
+			} else {
+				if ( 'ignore_role_' === substr( $option, 0, 12 ) ) {
```

Comment:
Will there be a field to control that option in Calypso? If not, maybe we can remove this?